File: matlab/toolbox/examples/export_png_images.m
function export_png_images(input, kwargs)
    arguments
        input (1,1) string
        kwargs.output_prefix (1,1) string = "image_"
        kwargs.show (1,1) logical = false
    end

    save = @imwrite;
    if kwargs.show
        save = @imwrite_and_show;
    end

    import mrd.binary.MrdReader
    r = MrdReader(input);

    header = r.read_header();
    if header == yardl.None
        error("No header found in input file");
    end

    image_counter = 0;
    while r.has_data()
        item = r.read_data();

        if ~item.isImageFloat()
            error("Stream must contain only floating point images");
        end

        image = item.value;

        image.data = image.data ./ max(image.data(:));

        if ismatrix(image.data)
            save(transpose(image.data), sprintf("%s%04d.png", kwargs.output_prefix, image_counter));
            image_counter = image_counter + 1;
        elseif ndims(image.data) == 3
            shape = size(image.data);
            for s = 1:shape(3)
                save(transpose(image.data(:,:,s)), sprintf("%s%04d.png", kwargs.output_prefix, image_counter));
                image_counter = image_counter + 1;
            end
        else
            for c = 1:image.channels()
                for s = 1:image.slices()
                    save(transpose(image.data(:,:,s,c)), sprintf("%s%04d.png", kwargs.output_prefix, image_counter));
                    image_counter = image_counter + 1;
                end
            end
        end
    end

    if kwargs.show
        % Wait for all figure windows to be closed
        uiwait;
    end
end

function imwrite_and_show(image_data, filename)
    figure;
    imshow(image_data);
    imwrite(image_data, filename);
end

File: python/mrd/tools/export_png_images.py
import sys
import argparse
import numpy as np
from PIL import Image

import mrd


def export(input, output, verbose=False):
    with mrd.BinaryMrdReader(input) as r:
        header = r.read_header()
        image_count = 0
        for item in r.read_data():
            if not isinstance(item, mrd.StreamItem.ImageFloat):
                raise RuntimeError("Stream must contain only floating point images")

            image = item.value
            image.data *= 255 / image.data.max()
            pixels = image.data.astype(np.uint8)

            for c in range(image.channels()):
                for s in range(image.slices()):
                    im = Image.fromarray(pixels[c, s, :, :], 'L')
                    filename = f"{output}{image_count:05d}.png"
                    im.save(filename, format='PNG')
                    if verbose:
                        print(f"Generated image {filename}")
                    image_count += 1


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Reconstructs an MRD stream")
    parser.add_argument('-i', '--input', type=str, required=False,
                        help="Input file (default stdin)")
    parser.add_argument('-o', '--output-prefix', type=str, required=False,
                        help="Output filename prefix", default="image_")
    parser.add_argument('-v', '--verbose', action='store_true', help="Print progress to stderr")
    args = parser.parse_args()

    input = open(args.input, "rb") if args.input is not None else sys.stdin.buffer

    export(input, args.output_prefix, args.verbose)
